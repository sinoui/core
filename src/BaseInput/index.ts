import BaseInput, { BaseInputProps } from './BaseInput';

export default BaseInput;

export type { BaseInputProps };
