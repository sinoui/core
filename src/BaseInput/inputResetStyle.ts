import { css } from 'styled-components';

/**
 * 输入框重置样式
 */
const inputResetStyle = css`
  display: block;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.5em;
  min-height: 1.5em;
  caret-color: ${({ theme }) => theme.palette.primary.main};

  ::-moz-placeholder {
    color: ${(props) => props.theme.palette.text.hint};
  }
  ::-webkit-input-placeholder {
    color: ${(props) => props.theme.palette.text.hint};
  }
  :-ms-input-placeholder {
    color: ${(props) => props.theme.palette.text.hint};
  }

  ::-ms-clear {
    display: none;
  }

  :focus {
    outline: 0;
  }

  :invalid {
    box-shadow: none;
    caret-color: ${({ theme }) => theme.palette.error.main};
  }
`;

export default inputResetStyle;
