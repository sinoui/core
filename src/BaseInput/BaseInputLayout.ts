import styled from 'styled-components';
import inputResetStyle from './inputResetStyle';

interface Props {
  /**
   * 如果设置为`true`，则输入框处于禁用状态。默认为`false`。
   */
  disabled?: boolean;
  /**
   * 如果设置为`true`，则输入框会占满宽度。默认为`false`。
   */
  $fullWidth?: boolean;
  /**
   * 输入框文本对齐方式
   */
  $align?: 'start' | 'end';
  /**
   * 是否是多行输入框
   */
  $multiline?: boolean;
  /**
   * 是否显示clear图标
   */
  $isShowClear?: boolean;
  /**
   * 是否有后缀元素
   */
  $hasEndAdornment?: boolean;
}

/**
 * 基础输入框的布局组件
 */
const BaseInputLayout = styled.div<Props>`
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  align-items: center;

  cursor: ${({ disabled }) => (disabled ? 'default' : 'text')};
  color: ${({ theme, disabled }) =>
    disabled ? theme.palette.text.disabled : theme.palette.text.primary};
  font-size: ${(props) => props.theme.typography.body1.fontSize};
  letter-spacing: ${(props) => props.theme.typography.body1.letterSpacing};
  font-family: ${(props) => props.theme.typography.fontFamily};
  line-height: 1.5em;

  ${({ $fullWidth }) => $fullWidth && 'width: 100%;'};

  > .sinoui-base-input__input {
    padding: 3.5px 0 4.5px;
    ${inputResetStyle};
    ${({ $align }) => $align === 'end' && 'text-align: right;'}
  }

  &.sinoui-base-input--multiline > textarea {
    resize: none;
    padding: 0;
  }

  .sinoui-base-input__clear {
    display: ${({ $hasEndAdornment }) => ($hasEndAdornment ? 'none' : 'flex')};
    font-size: 18px;
    cursor: pointer;
  }

  &:hover {
    .sinoui-input-adornment--end {
      display: ${({ $isShowClear }) => ($isShowClear ? 'none' : 'flex')};
    }

    .sinoui-base-input__clear {
      display: flex;
    }
  }

  ${({ $multiline }) => $multiline && 'padding: 3.5px 0 4.5px;'}
`;

export default BaseInputLayout;
