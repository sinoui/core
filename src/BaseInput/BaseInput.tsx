import React, { useCallback, useRef, useMemo, useEffect } from 'react';
import classNames from 'classnames';
import AutosizeTextarea from '@sinoui/core/AutosizeTextarea';
import useMultiRefs from '../utils/useMultiRefs';
import BaseInputLayout from './BaseInputLayout';
import mergeCallbacks from '../utils/mergeCallbacks';
import bemClassNames from '../utils/bemClassNames';
import { useFormControlContext } from '../FormControl';
import InputAdornment from '../InputAdornment';
import ClearIcon from './ClearIcon';

interface MultilineProps {
  /**
   * 指定多行输入框的最大行数
   */
  maxRows?: number;
  /**
   * 指定多行输入框的最小行数
   */
  minRows?: number;
  /**
   * 指定多行输入框的行数
   */
  rows?: number;
}

export interface BaseInputProps<
  InputComponentType extends React.ElementType = any,
  InputElementType = any,
  ExtendInputProps = {}
> {
  /**
   * 值
   */
  value?: string;
  /**
   * 默认值
   */
  defaultValue?: string;
  /**
   * 占位符文字
   */
  placeholder?: string;
  /**
   * 输入框类型
   */
  type?: string;
  /**
   * 输入框属性
   */
  inputProps?: React.ComponentPropsWithRef<InputComponentType> &
    Record<string, any> &
    ExtendInputProps;
  /**
   * 为输入框元素指定 ref
   */
  inputRef?: React.Ref<InputElementType>;
  /**
   * 是否是多行文本输入框。如果设置为`true`，则采用 react-textarea-autosize 渲染出 textarea
   */
  multiline?: boolean;
  /**
   * 是否只读，默认为false
   */
  readOnly?: boolean;
  /**
   * 必填
   */
  required?: boolean;
  /**
   * 自动补全
   */
  autoComplete?: string;
  /**
   * 自动获取焦点
   */
  autoFocus?: boolean;
  /**
   * 自定义样式类名称
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 失去焦点时的回调函数
   */
  onBlur?: (event: React.FocusEvent<InputElementType>) => void;
  /**
   * 获取焦点时的回调函数
   */
  onFocus?: (event: React.FocusEvent<InputElementType>) => void;
  /**
   * 值变化时的回调函数
   */
  onChange?: (event: React.ChangeEvent<InputElementType>) => void;
  /**
   * 点击事件的回调函数
   */
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  /**
   * 在组件的开头添加装饰器`InputAdornment`。
   */
  startAdornment?: React.ReactNode;
  /**
   * 在组件的末尾添加装饰器`InputAdornment`。
   */
  endAdornment?: React.ReactNode;
  /**
   * 不可用状态
   */
  disabled?: boolean;
  /**
   * 全宽模式
   */
  fullWidth?: boolean;
  /**
   * 指定包含的子元素
   */
  children?: React.ReactNode;
  /**
   * 输入区域渲染元素
   */
  inputComponent?: React.ElementType;
  /**
   * 输入框文本对齐方式。默认为`start`。
   */
  align?: 'start' | 'end';
  /**
   * 引用根元素
   */
  ref?: React.Ref<HTMLDivElement>;
  /**
   * 设置输入框元素的id
   */
  id?: string;
  /**
   * 设置输入框的name
   */
  name?: string;
  /**
   * 最小行数目
   */
  minRows?: number;
  /**
   * 最大行数目
   */
  maxRows?: number;
  /**
   * 指定输入框校验错误信息
   */
  error?: boolean;
  /**
   * 指定输入框校验错误信息
   */
  errorText?: string;
  /**
   * 是否允许清除
   */
  allowClear?: boolean;
  /**
   * 点击清除按钮时的回调函数
   */
  onClear?: (event: React.MouseEvent<HTMLOrSVGElement>) => void;
}

export interface BaseInputComponentType<InputElementType = HTMLInputElement> {
  <C extends React.ElementType>(
    props: {
      inputComponent: C;
    } & BaseInputProps<C, InputElementType>,
  ): JSX.Element | null;
  (
    props: {
      multiline: true;
    } & BaseInputProps<'textarea', HTMLTextAreaElement, MultilineProps>,
  ): JSX.Element | null;
  (props: BaseInputProps<'input', InputElementType>): JSX.Element | null;
}

/**
 * 基础输入框组件
 *
 * `BaseInput` 是一个包含尽量少样式的组件。它是用来简化构建输入框的基础组件。包含了输入框的重置样式和部分状态控制。
 */
const BaseInput: BaseInputComponentType = React.forwardRef<
  HTMLDivElement,
  BaseInputProps
>(function BaseInput(props, ref) {
  const formControlContext = useFormControlContext();
  const {
    value,
    type,
    inputProps: inputPropsProp = {},
    inputRef: inputRefProp,
    autoComplete,
    autoFocus,
    className,
    multiline,
    inputComponent = 'input',
    startAdornment,
    endAdornment,
    required,
    onChange,
    onBlur,
    onFocus,
    onClick,
    disabled,
    fullWidth,
    children,
    placeholder,
    align,
    id = formControlContext?.id,
    name,
    minRows,
    maxRows,
    error,
    errorText,
    allowClear,
    onClear,
    ...other
  } = props;

  const inputRef = useRef<any>(null);
  const handleInputRef = useMultiRefs(
    inputRef,
    inputRefProp,
    inputPropsProp.ref as any,
  );

  useEffect(() => {
    const input: HTMLInputElement = inputRef.current;
    if (input && error && input.setCustomValidity) {
      input.setCustomValidity(errorText ?? 'error');
    }
  }, [error, errorText]);

  const {
    onChange: onChangeInputProp,
    onBlur: onBlurInputProp,
    onFocus: onFocusInputProp,
  } = inputPropsProp;

  /**
   * 值变更时的回调函数
   */
  const handleChange = useMemo(
    () => mergeCallbacks(onChange, onChangeInputProp),
    [onChangeInputProp, onChange],
  );

  const onBlurFromContext = formControlContext?.onBlur;
  const onFocusFromContext = formControlContext?.onFocus;

  /**
   * 失去焦点时的回调函数
   */
  const handleBlur = useMemo(
    () => mergeCallbacks(onBlur, onBlurInputProp, onBlurFromContext),
    [onBlur, onBlurFromContext, onBlurInputProp],
  );

  /**
   * 获取焦点时的回调函数
   */
  const handleFocus = useMemo(
    () => mergeCallbacks(onFocus, onFocusInputProp, onFocusFromContext),
    [onFocus, onFocusFromContext, onFocusInputProp],
  );

  /**
   * 处理点击事件的回调函数xx
   */
  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      const input = inputRef.current;
      if (input && event.currentTarget === event.target) {
        input.focus();
      }
      if (onClick) {
        onClick(event);
      }
    },
    [onClick],
  );

  const handleClear = useCallback(
    (event: React.MouseEvent<HTMLOrSVGElement>) => {
      event.preventDefault();
      event.stopPropagation();
      if (onClear) {
        onClear(event);
        return;
      }
      event.persist();
      Object.defineProperty(event, 'target', {
        writable: true,
        value: { value: '' },
      });
      handleChange(event);
    },
    [handleChange, onClear],
  );

  const isShowClear =
    allowClear &&
    (Array.isArray(value) ? value.length > 0 : !!value) &&
    !disabled &&
    !props.readOnly;

  const InputComonent = multiline ? AutosizeTextarea : inputComponent;

  const inputprops: Record<string, any> = {
    ...inputPropsProp,
    value,
    id,
    name,
    type,
    autoFocus,
    autoComplete,
    disabled,
    readOnly: props.readOnly,
    defaultValue: props.defaultValue,
    onChange: handleChange,
    onBlur: handleBlur,
    onFocus: handleFocus,
    placeholder,
    'aria-required': required,
    ref: handleInputRef,
  };

  if (multiline) {
    inputprops.minRows = minRows;
    inputprops.maxRows = maxRows;
  }

  return (
    <BaseInputLayout
      className={bemClassNames(
        'sinoui-base-input',
        {
          disabled,
          multiline,
        },
        className,
      )}
      disabled={disabled}
      $fullWidth={fullWidth}
      $multiline={multiline}
      $align={align}
      $isShowClear={isShowClear}
      $hasEndAdornment={!!endAdornment}
      data-testid="baseInput"
      ref={ref}
      onClick={handleClick}
      {...other}
    >
      {startAdornment}
      <InputComonent
        {...inputprops}
        className={classNames(inputprops.className, 'sinoui-base-input__input')}
      />
      {isShowClear && (
        <InputAdornment position="end" className="sinoui-base-input__clear">
          <ClearIcon onClick={handleClear} />
        </InputAdornment>
      )}
      {endAdornment}
      {children}
    </BaseInputLayout>
  );
});

export default BaseInput;
