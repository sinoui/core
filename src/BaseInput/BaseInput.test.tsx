import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import BaseInput from './BaseInput';

it('正确渲染BaseInput', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <BaseInput />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染disabled属性下的BaseInput', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <BaseInput disabled />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染全宽模式下的BaseInput', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <BaseInput fullWidth />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染多行输入框', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <BaseInput multiline disabled />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('值改变时，onChange被调用', () => {
  const onChange = jest.fn();
  const { getByPlaceholderText } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput onChange={onChange} placeholder="请输入内容" />
    </ThemeProvider>,
  );
  const input = getByPlaceholderText('请输入内容');
  fireEvent.change(input, { target: { value: '123' } });

  expect(onChange).toHaveBeenCalledTimes(1);
});

it('值改变时，inputProps中的onChange被调用', () => {
  const onChange = jest.fn();
  const { getByPlaceholderText } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput inputProps={{ onChange }} placeholder="请输入内容" />
    </ThemeProvider>,
  );
  const input = getByPlaceholderText('请输入内容');
  fireEvent.change(input, { target: { value: '123' } });

  expect(onChange).toHaveBeenCalledTimes(1);
});

it('获取焦点时，onFocus事件被调用', () => {
  const onFocus = jest.fn();
  const inputOnFocus = jest.fn();
  const { getByPlaceholderText } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput
        onFocus={onFocus}
        inputProps={{ onFocus: inputOnFocus }}
        placeholder="请输入内容"
      />
    </ThemeProvider>,
  );
  const input = getByPlaceholderText('请输入内容');
  fireEvent.focus(input);
  expect(onFocus).toHaveBeenCalledTimes(1);
  expect(inputOnFocus).toHaveBeenCalled();
});

it('失去焦点时，onBlur被调用', () => {
  const onBlur = jest.fn();
  const inputOnblur = jest.fn();

  const { getByPlaceholderText } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput
        onBlur={onBlur}
        inputProps={{ onBlur: inputOnblur }}
        placeholder="请输入内容"
      />
    </ThemeProvider>,
  );
  const input = getByPlaceholderText('请输入内容');

  fireEvent.blur(input);

  expect(onBlur).toHaveBeenCalled();
  expect(inputOnblur).toHaveBeenCalled();
});

it('渲染修饰元素', () => {
  const { getByText } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput renderSuffix={(_state) => <div>这是修饰元素</div>} />
    </ThemeProvider>,
  );

  expect(getByText('这是修饰元素')).toBeDefined();
});

it('点击获取焦点', () => {
  const handleFocus = jest.fn();
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <BaseInput onFocus={handleFocus} />
    </ThemeProvider>,
  );

  act(() => {
    fireEvent.click(getByTestId('baseInput'));
  });

  expect(handleFocus).toBeCalled();
});
