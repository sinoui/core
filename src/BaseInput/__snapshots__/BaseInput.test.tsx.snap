// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正确渲染BaseInput 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  cursor: pointer;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.1875rem;
}

.c0 > input {
  padding: 6px 0 7px;
  border: 0px;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: rgba(0,0,0,0.87);
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  display: block;
  min-width: 0px;
  width: 100%;
  height: 1.1875rem;
}

.c0 > input::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear {
  display: none;
}

.c0 > input:focus {
  outline: 0;
}

.c0 > input:invalid {
  box-shadow: none;
}

.c0 > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__endcomponent {
  display: block;
}

<div
  className="c0 sinoui-base-input__layout"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`渲染disabled属性下的BaseInput 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  cursor: default;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.1875rem;
}

.c0 > input {
  padding: 6px 0 7px;
  border: 0px;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: rgba(0,0,0,0.38);
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  display: block;
  min-width: 0px;
  width: 100%;
  height: 1.1875rem;
  opacity: 1;
}

.c0 > input::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear {
  display: none;
}

.c0 > input:focus {
  outline: 0;
}

.c0 > input:invalid {
  box-shadow: none;
}

.c0 > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__endcomponent {
  display: block;
}

<div
  className="c0 sinoui-base-input__layout"
  data-testid="baseInput"
  disabled={true}
  onClick={[Function]}
>
  <input
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`渲染全宽模式下的BaseInput 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  cursor: pointer;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.1875rem;
  width: 100%;
}

.c0 > input {
  padding: 6px 0 7px;
  border: 0px;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: rgba(0,0,0,0.87);
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  display: block;
  min-width: 0px;
  width: 100%;
  height: 1.1875rem;
}

.c0 > input::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear {
  display: none;
}

.c0 > input:focus {
  outline: 0;
}

.c0 > input:invalid {
  box-shadow: none;
}

.c0 > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__endcomponent {
  display: block;
}

<div
  className="c0 sinoui-base-input__layout"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`渲染多行输入框 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  cursor: default;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  line-height: 1.1875rem;
}

.c0 > input {
  padding: 6px 0 7px;
  border: 0px;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: rgba(0,0,0,0.38);
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  display: block;
  min-width: 0px;
  width: 100%;
  height: 1.1875rem;
  opacity: 1;
}

.c0 > input::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear {
  display: none;
}

.c0 > input:focus {
  outline: 0;
}

.c0 > input:invalid {
  box-shadow: none;
}

.c0 > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__clear {
  display: block;
}

.c0:hover > .sinoui-base-input__endcomponent {
  display: block;
}

.c1 {
  border: 0px;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  color: rgba(0,0,0,0.38);
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  display: block;
  min-width: 0px;
  width: 100%;
  height: 1.1875rem;
  resize: none;
  overflow: auto;
  padding: 6px 0;
  opacity: 1;
}

.c1::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c1::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c1:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c1::-ms-clear {
  display: none;
}

.c1:focus {
  outline: 0;
}

.c1:invalid {
  box-shadow: none;
}

<div
  className="c0 sinoui-base-input__layout"
  data-testid="baseInput"
  disabled={true}
  onClick={[Function]}
>
  <textarea
    className="c1"
    disabled={true}
    multiline={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  />
</div>
`;
